GitHubReferences.ps1

```diff
@@ -0,0 +1,525 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+function Get-GitHubReference
+{
+<#
+    .SYNOPSIS
+        Retrieve a reference from a given GitHub repository.
+
+    .DESCRIPTION
+        Retrieve a reference from a given GitHub repository.
+        The Git repo for this module can be found here: http://aka.ms/PowerShellForGitHub
+
+    .PARAMETER OwnerName
+        Owner of the repository.
+        If not supplied here, the DefaultOwnerName configuration property value will be used.
+
+    .PARAMETER RepositoryName
+        Name of the repository.
+        If not supplied here, the DefaultRepositoryName configuration property value will be used.
+
+    .PARAMETER Uri
+        Uri for the repository.
+        The OwnerName and RepositoryName will be extracted from here instead of needing to provide
+        them individually.
+
+    .PARAMETER Tag
+        The name of the Tag to be retrieved.
+
+    .PARAMETER Branch
+        The name of the Branch to be retrieved.
+
+    .PARAMETER All
+        If provided, will retrieve both branches and tags for the given repository
+
+    .PARAMETER MatchPrefix
+        If provided, this will return matching preferences for the given branch/tag name in case an exact match is not found
+
+    .PARAMETER AccessToken
+        If provided, this will be used as the AccessToken for authentication with the
+        REST Api. Otherwise, will attempt to use the configured value or will run unauthenticated.
+
+    .PARAMETER NoStatus
+        If this switch is specified, long-running commands will run on the main thread
+        with no commandline status update.  When not specified, those commands run in
+        the background, enabling the command prompt to provide status information.
+        If not supplied here, the DefaultNoStatus configuration property value will be used.
+
+    .OUTPUTS
+        [PSCustomObject]
+        Details of the git reference in the given repository
+
+    .EXAMPLE
+        Get-GitHubReference -OwnerName microsoft -RepositoryName PowerShellForGitHub -Tag powershellTagV1
+
+    .EXAMPLE
+        Get-GitHubReference -OwnerName microsoft -RepositoryName PowerShellForGitHub -Branch master
+
+    .EXAMPLE
+        Get-GitHubReference -OwnerName microsoft -RepositoryName PowerShellForGitHub -Branch powershell -MatchPrefix
+
+        Get the branch 'powershell' and if it doesn't exist, get all branches beginning with 'powershell'
+
+    .EXAMPLE
+        Get-GitHubReference -Uri https://github.com/You/YourRepo -All

```

Comment:
I had to add an 'All' parameter because of the parametersets. Since both branch and tag are optional, calling Get-GithubReference without parameters results in PS not being able to find a parameterset, hence this workaround. Appreciate any suggestions on bettering this

Reply:
Thanks. I think this greatly simplifies it!